packages/flutter/lib/src/material/menu_bar.dart

```diff
@@ -0,0 +1,3396 @@
+    // edge of the screen by the button padding. If the menu is off of the screen,
+    // move the menu to the other side of the button first, and then if it
+    // doesn't fit there, then just move it over as much as needed to make it
+    // fit.
+    if (childSize.width >= allowedRect.width) {
+      // It just doesn't fit, so put as much on the screen as possible.
+      x = allowedRect.left;
+    } else {
+      if (offLeftSide(x)) {
+        // If the parent is a different orientation than the current one, then
+        // just push it over instead of trying the other side.
+        if (parentOrientation != orientation) {
+          x = allowedRect.left;
+        } else {
+          final double newX = absoluteButtonRect.right;
+          if (!offRightSide(newX)) {
+            x = newX;
+          } else {
+            x = allowedRect.left;
+          }
+        }
+      } else if (offRightSide(x)) {
+        if (parentOrientation != orientation) {
+          x = allowedRect.right - childSize.width;
+        } else {
+          final double newX = absoluteButtonRect.left - childSize.width;
+          if (!offLeftSide(newX)) {
+            x = newX;
+          } else {
+            x = allowedRect.right - childSize.width;
+          }
+        }
+      }
+    }
+    if (childSize.height >= allowedRect.height) {
+      // Too tall to fit, fit as much on as possible.
+      y = allowedRect.top;
+    } else {
+      if (offTop(y)) {
+        final double newY = absoluteButtonRect.bottom;
+        if (!offBottom(newY)) {
+          y = newY;
+        } else {
+          y = allowedRect.top;
+        }
+      } else if (offBottom(y)) {
+        final double newY = absoluteButtonRect.top - childSize.height;
+        if (!offTop(newY)) {
+          y = newY;
+        } else {
+          y = allowedRect.bottom - childSize.height;
+        }
+      }
+    }
+    return Offset(x, y);
+  }
+
+  Rect _closestScreen(Iterable<Rect> screens, Offset point) {
+    Rect closest = screens.first;
+    for (final Rect screen in screens) {
+      if ((screen.center - point).distance < (closest.center - point).distance) {
+        closest = screen;
+      }
+    }
+    return closest;
+  }
+
+  @override
+  bool shouldRelayout(_MenuLayout oldDelegate) {
+    return buttonRect != oldDelegate.buttonRect ||
+        textDirection != oldDelegate.textDirection ||
+        buttonPadding != oldDelegate.buttonPadding ||
+        alignment != oldDelegate.alignment ||
+        alignmentOffset != oldDelegate.alignmentOffset ||
+        globalMenuPosition != oldDelegate.globalMenuPosition ||
+        orientation != oldDelegate.orientation ||
+        parentOrientation != oldDelegate.parentOrientation ||
+        !setEquals(avoidBounds, oldDelegate.avoidBounds);
+  }
+}
+
+// Base class for all menu nodes that make up the menu tree, to allow walking of
+// the tree for navigation.
+abstract class _MenuNode with DiagnosticableTreeMixin, ChangeNotifier {
```

Comment:
The _MenuNode / MenuEntry split remains a bit suspect to me. It feels like these should bit one and the same thing with a public API to be used by anyone and a private API that's only used within this file.

After all, MenuController is also just one class and not split into two. It has a public API (e.g. `closeAll`) and a file-private API (e.g. `_menuOpened`)

Reply:
Yeah, OK.  I'm sure I can merge them.  I'll work on that.